### PR TITLE
Refactor SkyBox

### PR DESCRIPTION
**Pull request by gz65555**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)

Add a background property to scene which can set the sky and background color of scene.

### What is the current behavior? (You can also link to an open issue here)

You have to add a component to set sky box: 

``` TypeScript
scene.getRootEntity().addComponent(SkyBox).skyBoxMap = cubeMap;
```

And you have to set background color by camera:

``` TypeScript
camera.backgroundColor = new Vector4(1.0, 0.25, 0.25, 1.0);
```

And `camera.clearMode` api is redundant, which developer have never used.

### What is the new behavior (if this is a feature change)?

I have unioned background sky and background color togethor, which make background could be easily be understood. And we also can add texture background in the future.

``` TypeScript
// set background sky
scene.background = Sky.createSkybox(engine, cubeMap);
// set background color
scene.background = new Color(0.25, 0.25, 0.25, 1.0);
```

### Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)

Yes.

The `clearMode` property of Camera has been deleted.
The `backgroundColor` property of Camera has been deleted.
The Skybox component has been deleted.

Which use `Scene.background` instead.

### CODE REVIEW

File: packages/core/src/Scene.ts [hunk 1]
```diff
@@ -22,6 +23,8 @@ export class Scene extends EngineObject {
   /** scene-related shaderdata  */
   readonly shaderData: ShaderData = new ShaderData(ShaderDataGroup.Scene);
 
+  /** The background of the scene, which could be a sky or a color. */
+  background: Sky | Color = new Color(0.25, 0.25, 0.25, 1)
```
Comment (GuoLei1990) [inline packages/core/src/Scene.ts:27]: use backgroundMode design，default can be sky.

Comment (GuoLei1990) [reply packages/core/src/Scene.ts:27]: Design reference
```typescript
class Scene
{
   ...
   readonly background: SceneBackground = new SceneBackground();
   ...
}
```
```typescript
enum BackgroundColorMode
{
    Solid,
    Texture,// todo:you can to do later
    Sky
}
```

```typescript
class SceneBackground
{
    clearColor: boolean=true;
    clearDepth: boolean=true;
    clearStencil: boolean=true;

    colorMode: BackgroundColorMode = BackgroundColorMode.Solid;

    colorSolid: Color = new Color(0.25,0.25,0.25,1.0);    
    colorTexture: Texture2D;// todo:you can to do later
    readonly colorSky: Sky = new Sky();
}
```
NOTE:If `colorTexture` and `colorSky` not invaild, for example: `null`, will fall back to solid. 

Usage:
```typeScript
scene.background.colorMode = BackgroundColorMode.Sky;
const sky = scene.background.colorSky;
sky.mesh = PrimitiveMesh.createCube();
sky.material = new AtmosphericSkyMaterial();
```

```typeScript
scene.background.colorMode = BackgroundColorMode.Texture;
scene.background.colorTexture = texture;
```

```typeScript
scene.background.colorMode = BackgroundColorMode.Solid;
scene.background.colorSolid =  new color(1.0,0.0,0.0);
```

**Comment by GuoLei1990** [comment]
**Design reference,here is some points:**
- Background in scene is very natural, background mode has sky、texture、solidColor, etc...
- But how to clear the color, depth and stencil. is the ability of the camera (We should keep), the camera that renders the scene has the choice whether to clear the depth or not clear nothing,etc.

```typescript
enum BackgroundMode
{
    /* Solid color. */
    Solid,
    /* Sky. */
    Sky,
    /* Texture. */
    Texture  // todo: you can to do later
}
```
```typescript
/**
 * Determining what to clear when rendering by a Camera.
 */
export enum CameraClearFlags {
  /* Clear depth and color from background. */
  DepthColor,
  /* Clear depth only. */
  Depth,
  /* Do nothing. */
  None
}
```

```typescript
class Scene
{
   ...
   /* Scene background. */
   readonly background: Background = new Background();
   ...
}
```

```typescript
class Background
{
    mode: BackgroundColorMode = BackgroundMode.Solid;
    solid: Color = new Color(0.25,0.25,0.25,1.0);    
    readonly sky: Sky = new Sky();x
    texture: Texture2D;  // todo: you can to do later
}
```

Note: If `texture` or `sky` not invaild (for example: texture is `null`), will fall back to solid even mode is texture or sky.  

**Usage**
```typeScript
scene.background.mode = BackgroundMode.Sky;
const sky = scene.background.sky;
sky.mesh = PrimitiveMesh.createCube();
sky.material = new AtmosphericSkyMaterial();
```

```typeScript
scene.background.mode = BackgroundMode.Texture;
scene.background.texture = texture;
```

```typeScript
scene.background.mode = BackgroundMode.Solid;
scene.background.solid =  new color(1.0,0.0,0.0);
```

**Comment by gz65555** [comment]
`solid` is not  comprehensible for front-end engineer, which means solid line for border. I think `color` would be better.

**Comment by gz65555** [comment]
And I will not add `Texture` mode on this pull request. Maybe next time.

File: packages/core/src/sky/Sky.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import { SkyBoxMaterial } from "./SkyBoxMaterial";
+
+export class Sky {
+  /**
+   * Create a sky box.
+   * @param engine - current engine
+   * @param textureCubeMap - cube map texture of the sky
+   * @returns 
+   */
+  static createSkybox(engine: Engine, textureCubeMap: TextureCubeMap): Sky {
```
[the remark this answers is not in the document]
Comment (gz65555) [reply packages/core/src/sky/Sky.ts]: +1

File: packages/core/src/Scene.ts (hunk 1 above)
Comment (gz65555) [reply packages/core/src/Scene.ts]: +1

File: packages/core/src/sky/Sky.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { Matrix } from "@oasis-engine/math";
+import { Engine } from "../Engine";
+import { Mesh } from "../graphic";
+import { Material } from "../material";
+import { PrimitiveMesh } from "../mesh/PrimitiveMesh";
```
Comment (GuoLei1990) [inline packages/core/src/sky/Sky.ts:5]: Clear imports

File: packages/core/src/Background.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import { Color } from "@oasis-engine/math";
+import { Sky } from "./sky/Sky";
+
+/**
+ * The Background mode enumeration.
+ */
+export enum BackgroundMode {
```
Comment (GuoLei1990) [inline packages/core/src/Background.ts:7]: Use seperate file , for example `*/enums/BackgroundMode.ts` path.

File: packages/core/src/Background.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Background of scene.
+ */
+export class Background {
+  /**
+   * Background mode.
+   * @defaultValue `BackgroundMode.Sky`
+   */
+  mode: BackgroundMode = BackgroundMode.Sky;
+  /**
```
Comment (GuoLei1990) [inline packages/core/src/Background.ts:23]: Leave a blank line between multi-line comments.

File: packages/core/src/Camera.ts [hunk 6]
```diff
@@ -17,9 +16,16 @@ import { Transform } from "./Transform";
 import { UpdateFlag } from "./UpdateFlag";
 
 /**
- * @todo
+ * Camera clear flags enumeration.
  */
-type Sky = {};
+export enum CameraClearFlags {
```
Comment (GuoLei1990) [inline packages/core/src/Camera.ts:21]: Use seperate file , for example `*/enums/CameraClearFlags.ts` path.

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts [hunk 7]
```diff
@@ -140,17 +145,25 @@ export class BasicRenderPipeline {
...
+      if (clearFlags !== CameraClearFlags.None) {
+        rhi.clearRenderTarget(camera.engine, clearFlags, background.color);
+      }
 
       if (pass.renderOverride) {
         pass.render(camera, this._opaqueQueue, this._alphaTestQueue, this._transparentQueue);
       } else {
         this._opaqueQueue.render(camera, pass.replaceMaterial, pass.mask);
         this._alphaTestQueue.render(camera, pass.replaceMaterial, pass.mask);
+        if (background.mode === BackgroundMode.Sky && background.sky) {
```
Comment (GuoLei1990) [inline packages/core/src/RenderPipeline/BasicRenderPipeline.ts:164]: `background.sky` is always have, you should check `mesh` and `material` are both OK！
for example add internal fnction to `Sky`:
```typescript
_isVaild():booean
{
  return material&&mesh;
}
```

File: packages/rhi-webgl/src/WebGLRenderer.ts [hunk 8]
```diff
@@ -188,53 +188,41 @@ export class WebGLRenderer implements IHardwareRenderer {
...
-        break;
-    }
-
-    if (clearFlag & gl.COLOR_BUFFER_BIT && targetBlendState.colorWriteMask !== ColorWriteMask.All) {
-      gl.colorMask(true, true, true, true);
-      targetBlendState.colorWriteMask = ColorWriteMask.All;
+    let clearFlag = gl.DEPTH_BUFFER_BIT | gl.STENCIL_BUFFER_BIT;
+    if (clearFlags === CameraClearFlags.DepthColor) {
+      clearFlag = clearFlag | gl.COLOR_BUFFER_BIT;
+      if (clearColor.toGamma) {
```
Comment (GuoLei1990) [inline packages/rhi-webgl/src/WebGLRenderer.ts:206]: `if (clearColor.toGamma)` no useful, clearColor type is always `Color`

File: packages/core/src/Background.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+   * Background color.
```
Comment (GuoLei1990) [inline packages/core/src/Background.ts:24]: `Background solid color.`

File: packages/core/src/Background.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+   * @defaultValue `new Color(0.25, 0.25, 0.25, 1.0)`
+   */
+  color: Color = new Color(0.25, 0.25, 0.25, 1.0);
```
Comment (GuoLei1990) [inline packages/core/src/Background.ts:27]: add a tip: 

@remarks Take effect when mode is 'BackgroundMode.Sky'.

File: packages/core/src/Background.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  /*
+   * Background sky.
```
Comment (GuoLei1990) [inline packages/core/src/Background.ts:29]: add a tip:
@remarks Take effect when mode is 'BackgroundMode.Color'.

**Comment by GuoLei1990** [comment]
> `solid` is not comprehensible for front-end engineer, which means solid line for border. I think `color` would be better.

You are right! Solid is not good!

I think background have three mode `solid color`, `texture` and `sky`. 

The reason why it was not called `Color` before is because they are all modes of how to draw the background color.
And our camera's clearFlags is `CameraClearFlags.DepthColor`, so if we call that 'Color' is easy to misunderstand. Developers will think that the Color part of CameraClearFlags.DepthColor is only equal to 'solid color'.

But `Solid` is really not good, we should call it  'SolidColor'. What do you  think?

File: packages/rhi-webgl/src/WebGLRenderer.ts (hunk 8 above)
Comment (gz65555) [reply packages/rhi-webgl/src/WebGLRenderer.ts]: +1

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts [hunk 12]
```diff
@@ -176,4 +189,36 @@ export class BasicRenderPipeline {
       this._opaqueQueue.pushPrimitive(element);
     }
   }
+
+  private _drawSky(engine: Engine, camera: Camera, sky: Sky) {
```
[the remark this answers is not in the document]
Comment (gz65555) [reply packages/core/src/RenderPipeline/BasicRenderPipeline.ts]: +1

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts [hunk 13]
```diff
@@ -176,4 +189,36 @@ export class BasicRenderPipeline {
...
+    const { material, mesh, _matrix } = sky;
+    if (!material) {
+      Logger.warn("The material of sky is not defined.");
+      return;
+    }
+    if (!mesh) {
+      Logger.warn("The mesh of sky is not defined.");
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (gz65555) [reply packages/core/src/RenderPipeline/BasicRenderPipeline.ts:203]: +1

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts (hunk 7 above)
Comment (gz65555) [reply packages/core/src/RenderPipeline/BasicRenderPipeline.ts]: +1

File: packages/core/src/Camera.ts (hunk 6 above)
Comment (gz65555) [reply packages/core/src/Camera.ts]: +1

File: packages/core/src/Background.ts (hunk 5 above)
Comment (gz65555) [reply packages/core/src/Background.ts:15]: +1

File: packages/core/src/Background.ts (hunk 4 above)
Comment (gz65555) [reply packages/core/src/Background.ts]: +1

File: packages/core/src/sky/Sky.ts (hunk 3 above)
Comment (gz65555) [reply packages/core/src/sky/Sky.ts]: +1

File: packages/core/src/sky/Sky.ts [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+import { TextureCubeMap } from "../texture";
+import { SkyBoxMaterial } from "./SkyBoxMaterial";
+
+export class Sky {
```
Comment (GuoLei1990) [inline packages/core/src/sky/Sky.ts:5]: Add class comments.

**Review by GuoLei1990 — APPROVED** [review]
+1

**Pull request merged**